### PR TITLE
executor: support spill intermediate data for unparalleled hash agg

### PR DESCRIPTION
**Pull request by wshwsh12**
### What problem does this PR solve?

Issue Number: close #xxx 

Problem Summary: 

### What is changed and how it works?

Proposal: [Design](https://github.com/pingcap/tidb/pull/25792) 

What's Changed:
Based PR https://github.com/pingcap/tidb/pull/25820, introduce soft limit.
Support spilling intermediate date for unparalleled hashAgg.

How it Works:
1. When the memory usage is higher than the quota, switch the HashAgg executor to spill-mode.
2. When HashAgg is in spill-mode, keep the tuple in the hashMap no longer growing.
 a. If the key exists in the Map, aggreagte the result.
 b. If the key doesn't exist in the Map, spill the data to disk.
3. After all data have been processed,  output the aggreagte result in the Map, clear the Map. Then read the spilling data from disk, repeat the Step1-Step3 until all data have been aggregated.  

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`:

### Check List 

Tests 

- Unit test
- Integration test
- Manual test (add detailed scripts or steps below)
Set AggregateConcurrency to 1 and run all correctness test in tidb repo.

Side effects

- Performance regression
    - Consumes more CPU
    - Consumes more MEM

### Release note 

- Support spill intermediate data for unparalleled hash agg

### CODE REVIEW

**Comment by wshwsh12** [comment]
### Simple Benchmark
cpu: AMD Ryzen 7 3700X 8-Core Processor   
```
Old Implement:
BenchmarkHashAggRows
BenchmarkHashAggRows/(execType:hash,_aggFunc:sum,_ndv:1000,_hasDistinct:false,_rows:100000,_concurrency:1,_sorted:true)
BenchmarkHashAggRows/(execType:hash,_aggFunc:sum,_ndv:1000,_hasDistinct:false,_rows:100000,_concurrency:1,_sorted:true)-16         	      79	  23354686 ns/op
BenchmarkHashAggRows/(execType:hash,_aggFunc:sum,_ndv:1000,_hasDistinct:false,_rows:1000000,_concurrency:1,_sorted:true)
BenchmarkHashAggRows/(execType:hash,_aggFunc:sum,_ndv:1000,_hasDistinct:false,_rows:1000000,_concurrency:1,_sorted:true)-16        	       7	 150919829 ns/op
BenchmarkHashAggRows/(execType:hash,_aggFunc:sum,_ndv:1000,_hasDistinct:false,_rows:10000000,_concurrency:1,_sorted:true)
BenchmarkHashAggRows/(execType:hash,_aggFunc:sum,_ndv:1000,_hasDistinct:false,_rows:10000000,_concurrency:1,_sorted:true)-16       	       1	1315785257 ns/op

New Implement:
BenchmarkHashAggRows
BenchmarkHashAggRows/(execType:hash,_aggFunc:sum,_ndv:1000,_hasDistinct:false,_rows:100000,_concurrency:1,_sorted:true)
BenchmarkHashAggRows/(execType:hash,_aggFunc:sum,_ndv:1000,_hasDistinct:false,_rows:100000,_concurrency:1,_sorted:true)-16         	      52	  20963917 ns/op
BenchmarkHashAggRows/(execType:hash,_aggFunc:sum,_ndv:1000,_hasDistinct:false,_rows:1000000,_concurrency:1,_sorted:true)
BenchmarkHashAggRows/(execType:hash,_aggFunc:sum,_ndv:1000,_hasDistinct:false,_rows:1000000,_concurrency:1,_sorted:true)-16        	       7	 145583208 ns/op
BenchmarkHashAggRows/(execType:hash,_aggFunc:sum,_ndv:1000,_hasDistinct:false,_rows:10000000,_concurrency:1,_sorted:true)
BenchmarkHashAggRows/(execType:hash,_aggFunc:sum,_ndv:1000,_hasDistinct:false,_rows:10000000,_concurrency:1,_sorted:true)-16       	       1	1327683063 ns/op
```
### Memory usage test
Workload: tpch-sf=3g  1tidb 1tikv 1pd 
Config: oom-action = "log",mem-quota-query = 1 << 30, and close copr-cache
User-veriables: tidb_hashagg_final_concurrency = 1; tidb_hashagg_partial_concurrency = 1;

```
Old Implement
[tidb]> desc analyze select avg(L_EXTENDEDPRICE) from lineitem group by L_PARTKEY   ,L_SUPPKEY , L_QUANTITY;
+---------------------------+-------------+----------+-----------+----------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+----------+------+
| id                        | estRows     | actRows  | task      | access object  | execution info                                                                                                                                                                                                                                                                             | operator info                                                                                                                                                                                   | memory   | disk |
+---------------------------+-------------+----------+-----------+----------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+----------+------+
| HashAgg_9                 | 14397287.20 | 16776228 | root      |                | time:4m23.7s, loops:16385                                                                                                                                                                                                                                                                  | group by:test.lineitem.l_partkey, test.lineitem.l_quantity, test.lineitem.l_suppkey, funcs:avg(Column#19, Column#20)->Column#18                                                                 | 3.14 GB  | N/A  |
| └─TableReader_10          | 14397287.20 | 17963641 | root      |                | time:3m40.2s, loops:41, cop_task: {num: 40, max: 32.5s, min: 1.55s, avg: 19.7s, p95: 31.7s, max_proc_keys: 613883, p95_proc_keys: 462536, tot_proc: 12m46.5s, tot_wait: 9.29s, rpc_num: 40, rpc_time: 13m6.7s, copr_cache: disabled}                                                       | data:HashAgg_5                                                                                                                                                                                  | 320.5 MB | N/A  |
|   └─HashAgg_5             | 14397287.20 | 17963641 | cop[tikv] |                | tikv_task:{proc max:19.4s, min:873ms, p80:18.1s, p95:19.3s, iters:17591, tasks:40}, scan_detail: {total_process_keys: 17996609, total_keys: 17996649, rocksdb: {delete_skipped_count: 0, key_skipped_count: 17996609, block: {cache_hit_count: 57170, read_count: 0, read_byte: 0 Bytes}}} | group by:test.lineitem.l_partkey, test.lineitem.l_quantity, test.lineitem.l_suppkey, funcs:count(test.lineitem.l_extendedprice)->Column#19, funcs:sum(test.lineitem.l_extendedprice)->Column#20 | N/A      | N/A  |
|     └─TableFullScan_8     | 17996609.00 | 17996609 | cop[tikv] | table:lineitem | tikv_task:{proc max:10.9s, min:511ms, p80:10.1s, p95:10.7s, iters:17591, tasks:40}                                                                                                                                                                                                         | keep order:false, stats:pseudo                                                                                                                                                                  | N/A      | N/A  |
+---------------------------+-------------+----------+-----------+----------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+----------+------+
4 rows in set (4 min 24.836 sec)


New Implement:
[tidb]> desc analyze select avg(L_EXTENDEDPRICE) from lineitem group by L_PARTKEY   ,L_SUPPKEY , L_QUANTITY;
+---------------------------+-------------+----------+-----------+----------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+----------+---------+
| id                        | estRows     | actRows  | task      | access object  | execution info                                                                                                                                                                                                                                                                             | operator info                                                                                                                                                                                   | memory   | disk    |
+---------------------------+-------------+----------+-----------+----------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+----------+---------+
| HashAgg_9                 | 14397287.20 | 16776228 | root      |                | time:3m42.7s, loops:16385                                                                                                                                                                                                                                                                  | group by:test.lineitem.l_partkey, test.lineitem.l_quantity, test.lineitem.l_suppkey, funcs:avg(Column#19, Column#20)->Column#18                                                                 | 1.04 GB  | 2.80 GB |
| └─TableReader_10          | 14397287.20 | 17963641 | root      |                | time:58.2s, loops:41, cop_task: {num: 40, max: 35.4s, min: 1.64s, avg: 24s, p95: 33.5s, max_proc_keys: 613883, p95_proc_keys: 462536, tot_proc: 15m29.7s, tot_wait: 19.1s, rpc_num: 40, rpc_time: 16m0s, copr_cache: disabled}                                                             | data:HashAgg_5                                                                                                                                                                                  | 183.1 MB | N/A     |
|   └─HashAgg_5             | 14397287.20 | 17963641 | cop[tikv] |                | tikv_task:{proc max:19.5s, min:1.29s, p80:18.1s, p95:19.3s, iters:17591, tasks:40}, scan_detail: {total_process_keys: 17996609, total_keys: 17996649, rocksdb: {delete_skipped_count: 0, key_skipped_count: 17996609, block: {cache_hit_count: 57170, read_count: 0, read_byte: 0 Bytes}}} | group by:test.lineitem.l_partkey, test.lineitem.l_quantity, test.lineitem.l_suppkey, funcs:count(test.lineitem.l_extendedprice)->Column#19, funcs:sum(test.lineitem.l_extendedprice)->Column#20 | N/A      | N/A     |
|     └─TableFullScan_8     | 17996609.00 | 17996609 | cop[tikv] | table:lineitem | tikv_task:{proc max:11s, min:731ms, p80:10.1s, p95:10.8s, iters:17591, tasks:40}                                                                                                                                                                                                           | keep order:false, stats:pseudo                                                                                                                                                                  | N/A      | N/A     |
+---------------------------+-------------+----------+-----------+----------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+----------+---------+
4 rows in set (3 min 43.035 sec)
```

File: executor/aggregate.go [hunk 1]
```diff
@@ -939,6 +1002,29 @@ func (e *HashAggExec) execute(ctx context.Context) (err error) {
...
+			return nil
+		}
+	}
+	if e.processIdx < e.lastChunkNum {
+		e.childResult, err = e.listInDisk.GetChunk(e.processIdx)
+		if err != nil {
+			return err
+		}
+		e.processIdx++
+		return nil
```
Comment (mmyj) [inline executor/aggregate.go:1023]: lgtm. One picky problem, this line is useless. ^_^

**Review by mmyj — APPROVED** [review]

File: executor/aggregate.go [hunk 2]
```diff
@@ -1744,3 +1830,46 @@ func (e *vecGroupChecker) reset() {
...
+func (e *HashAggExec) ActionSpill() *AggSpillDiskAction {
+	if e.spillAction == nil {
+		e.spillAction = &AggSpillDiskAction{
+			e: e,
+		}
+	}
+	return e.spillAction
+}
+
+const maxSpillTimes = 10
```
Comment (XuHuaiyu) [inline executor/aggregate.go:1844]: Add a comment for this

File: executor/aggregate.go [hunk 3]
```diff
@@ -191,9 +192,18 @@ type HashAggExec struct {
...
 
 	stats *HashAggRuntimeStats
+
+	listInDisk   *chunk.ListInDisk
+	lastChunkNum int
+	processIdx   int
+	spillMode    uint32
+	spillChunk   *chunk.Chunk
+	spillAction  *AggSpillDiskAction
+	childDrained bool
```
Comment (XuHuaiyu) [inline executor/aggregate.go:206]: We need comments for these variables

File: executor/aggregate.go [hunk 4]
```diff
@@ -233,6 +243,15 @@ func (e *HashAggExec) Close() error {
 		if e.memTracker != nil {
 			e.memTracker.ReplaceBytesUsed(0)
 		}
+		if e.listInDisk != nil {
+			if err := e.listInDisk.Close(); err != nil {
+				return err
+			}
+		}
+		if e.spillAction != nil {
```
Comment (XuHuaiyu) [inline executor/aggregate.go:251]: What's this for?

File: executor/aggregate.go [hunk 5]
```diff
@@ -233,6 +243,15 @@ func (e *HashAggExec) Close() error {
 		if e.memTracker != nil {
 			e.memTracker.ReplaceBytesUsed(0)
 		}
+		if e.listInDisk != nil {
+			if err := e.listInDisk.Close(); err != nil {
+				return err
```
Comment (XuHuaiyu) [inline executor/aggregate.go:248]: Don't we close the chilrenExec? This may cause leaks.

File: executor/aggregate.go [hunk 6]
```diff
@@ -301,6 +320,17 @@ func (e *HashAggExec) initForUnparallelExec() {
 	e.groupKeyBuffer = make([][]byte, 0, 8)
 	e.childResult = newFirstChunk(e.children[0])
 	e.memTracker.Consume(e.childResult.MemoryUsage())
+
+	e.processIdx, e.lastChunkNum = 0, 0
+	e.executed, e.childDrained = false, false
+	e.listInDisk = chunk.NewListInDisk(retTypes(e.children[0]))
+	e.spillChunk = newFirstChunk(e.children[0])
+	if e.ctx.GetSessionVars().TrackAggregateMemoryUsage {
```
Comment (XuHuaiyu) [inline executor/aggregate.go:328]: Why do we need to check this?

File: executor/aggregate.go [hunk 7]
```diff
@@ -922,6 +974,17 @@ func (e *HashAggExec) execute(ctx context.Context) (err error) {
 		for j := 0; j < e.childResult.NumRows(); j++ {
 			groupKey := string(e.groupKeyBuffer[j]) // do memory copy here, because e.groupKeyBuffer may be reused.
 			if !e.groupSet.Exist(groupKey) {
+				if atomic.LoadUint32(&e.spillMode) == 1 && e.groupSet.Count() > 0 {
+					e.spillChunk.Append(e.childResult, j, j+1)
```
Comment (XuHuaiyu) [inline executor/aggregate.go:978]: Can we use `Chunk.sel` to optimize this if-block?
1. We can check `e.groupSet.Exist(groupKey)` and build the `sel` firstly, and then invoke `e.spillChunk.Append` based on the `sel`.
2. Further, if `len(sel) == len(e.childResult)`, we can invoke `e.listInDisk.Add(e.childResult)` directly.

File: executor/aggregate.go (hunk 4 above)
Comment (wshwsh12) [reply executor/aggregate.go]: Useless.. I remove the code now.

File: executor/aggregate.go (hunk 5 above)
Comment (wshwsh12) [reply executor/aggregate.go]: Addressed.

File: executor/aggregate.go (hunk 6 above)
Comment (wshwsh12) [reply executor/aggregate.go]: If tidb doesn't track aggregate executor memory usgae, should we also try to spill hashAgg when exceeded?
In addition, oom-use-tmp-storage also should be check... I add the check now. PTAL again.

File: executor/aggregate.go (hunk 7 above)
Comment (wshwsh12) [reply executor/aggregate.go]: Addressed.

File: executor/aggregate.go [hunk 8]
```diff
@@ -191,9 +193,18 @@ type HashAggExec struct {
...
 
-	memTracker *memory.Tracker // track memory usage.
+	memTracker  *memory.Tracker // track memory usage.
+	diskTracker *disk.Tracker
 
 	stats *HashAggRuntimeStats
+
+	listInDisk   *chunk.ListInDisk   // listInDisk is the chunks to store row values for spilling data.
+	lastChunkNum int                 // lastChunkNum indicates the num of spilling chunk.
+	processIdx   int                 // processIdx indicates the num of processed chunk in disk.
```
Comment (XuHuaiyu) [inline executor/aggregate.go:203]: offsetOfSpillChks

Comment (XuHuaiyu) [reply executor/aggregate.go:203]: This comment is hard to understand, we need a detailed comment to explain the usage of this var

File: executor/aggregate.go [hunk 9]
```diff
@@ -191,9 +193,18 @@ type HashAggExec struct {
...
+	spillMode    uint32              // spillMode means that no new groups are added to hash table.
```
Comment (XuHuaiyu) [inline executor/aggregate.go:204]: 1. isSpillModeSet?
2. Add an explanation for what does `0` and `1` mean

File: executor/aggregate.go [hunk 10]
```diff
@@ -191,9 +193,18 @@ type HashAggExec struct {
...
+	spillChunk   *chunk.Chunk        // spillChunk is the temp chunk for spilling.
```
Comment (XuHuaiyu) [inline executor/aggregate.go:205]: tmpChkForSpill

File: executor/aggregate.go (hunk 8 above)
Comment (wshwsh12) [reply executor/aggregate.go]: Addressed.

File: executor/aggregate.go [hunk 11]
```diff
@@ -934,11 +991,63 @@ func (e *HashAggExec) execute(ctx context.Context) (err error) {
 				allMemDelta += memDelta
 			}
 		}
+
+		// spill unprocessed data when exceeded.
+		if len(sel) > 0 {
+			err = e.spillUnprocessedData(sel)
```
[the remark this answers is not in the document]
Comment (wshwsh12) [reply executor/aggregate.go]: `e.childResult.SetSel(sel)` will let `len(sel) == len(e.childResult)` always true, and `e.listInDisk.Add(e.childResult)` directly. If there are only a few elements in sel, it maybe have performance issue.
I remove the logic `e.listInDisk.Add(e.childResult)` and always append to tmpChkForSpill, PTAL

File: executor/aggregate.go [hunk 12]
```diff
@@ -191,9 +193,30 @@ type HashAggExec struct {
...
-	memTracker *memory.Tracker // track memory usage.
+	memTracker  *memory.Tracker // track memory usage.
+	diskTracker *disk.Tracker
 
 	stats *HashAggRuntimeStats
+
+	// listInDisk is the chunks to store row values for spilling data.
+	// The HashAggExec may enter `spill mode` multiple times, and all spill data will append to ListInDisk.
+	listInDisk *chunk.ListInDisk
+	// numOfSpilledChks indicates the num of spilling chunk after the last round of processing is over.
```
Comment (XuHuaiyu) [inline executor/aggregate.go:204]:
```suggestion
	// numOfSpilledChks indicates the number of all the spilled chunks.
```

File: executor/aggregate.go [hunk 13]
```diff
@@ -191,9 +193,30 @@ type HashAggExec struct {
...
+	// After one round of processing is over, no data spilling again means that all data has been processed.
+	numOfSpilledChks int
+	// offsetOfSpillChks indicates the num of processed chunk in disk.
```
Comment (XuHuaiyu) [inline executor/aggregate.go:207]:
```suggestion
	// offsetOfSpillChks indicates the offset of the chunk be read from the disk.
```

File: executor/aggregate.go [hunk 14]
```diff
@@ -191,9 +193,30 @@ type HashAggExec struct {
...
+	// In the one round of processing, we need process all data spilled in the last round.
+	offsetOfSpillChks int
+
+	// inSpillMode indicates whether HashAgg is in `spill mode`.
+	// When HashAgg is in `spill mode`, keep the tuple in partialResultMap no longer growing.
```
Comment (XuHuaiyu) [inline executor/aggregate.go:212]:
```suggestion
	// When HashAgg is in `spill mode`, the size of `partialResultMap` is no longer growing and all the data fetched from the child executor is spilled to the disk.
```

File: executor/aggregate.go [hunk 15]
```diff
@@ -191,9 +193,30 @@ type HashAggExec struct {
...
+	memTracker  *memory.Tracker // track memory usage.
+	diskTracker *disk.Tracker
 
 	stats *HashAggRuntimeStats
+
+	// listInDisk is the chunks to store row values for spilling data.
+	// The HashAggExec may enter `spill mode` multiple times, and all spill data will append to ListInDisk.
+	listInDisk *chunk.ListInDisk
+	// numOfSpilledChks indicates the num of spilling chunk after the last round of processing is over.
+	// After one round of processing is over, no data spilling again means that all data has been processed.
```
Comment (XuHuaiyu) [inline executor/aggregate.go:205]:
```suggestion
```

File: executor/aggregate.go [hunk 16]
```diff
@@ -191,9 +193,30 @@ type HashAggExec struct {
...
+	numOfSpilledChks int
+	// offsetOfSpillChks indicates the num of processed chunk in disk.
+	// In the one round of processing, we need process all data spilled in the last round.
```
Comment (XuHuaiyu) [inline executor/aggregate.go:208]:
```suggestion
	// In each round of processing, we need to re-fetch all the chunks spilled in the last one.
```

File: executor/aggregate.go [hunk 17]
```diff
@@ -934,11 +1003,57 @@ func (e *HashAggExec) execute(ctx context.Context) (err error) {
...
+				return err
+			}
+		}
+
 		failpoint.Inject("ConsumeRandomPanic", nil)
 		e.memTracker.Consume(allMemDelta)
 	}
 }
 
+func (e *HashAggExec) spillUnprocessedData() (err error) {
```
Comment (XuHuaiyu) [inline executor/aggregate.go:1021]:
```suggestion
func (e *HashAggExec) spillUnprocessedData(isFullChk bool) (err error) {
if isFullChk {
return e.listInDisk.Add(e.childResult)
}
```

File: executor/aggregate.go [hunk 18]
```diff
@@ -934,11 +1003,57 @@ func (e *HashAggExec) execute(ctx context.Context) (err error) {
 				allMemDelta += memDelta
 			}
 		}
+
+		// spill unprocessed data when exceeded.
+		if len(sel) > 0 {
+			e.childResult.SetSel(sel)
+			err = e.spillUnprocessedData()
```
Comment (XuHuaiyu) [inline executor/aggregate.go:1010]:
```suggestion
			err = e.spillUnprocessedData(len(sel) == cap(sel))
```

File: executor/aggregate.go [hunk 19]
```diff
@@ -191,9 +193,30 @@ type HashAggExec struct {
...
+
+	// listInDisk is the chunks to store row values for spilled data.
+	// The HashAggExec may be set to `spill mode` multiple times, and all spilled data will be appended to ListInDisk.
+	listInDisk *chunk.ListInDisk
+	// numOfSpilledChks indicates the number of all the spilled chunks.
+	numOfSpilledChks int
+	// offsetOfSpilledChks indicates the offset of the chunk be read from the disk.
+	// In each round of processing, we need to re-fetch all the chunks spilled in the last one.
+	offsetOfSpilledChks int
+
```
Comment (XuHuaiyu) [inline executor/aggregate.go:209]:
```suggestion
```

**Review by XuHuaiyu — APPROVED** [review]
LGTM

**Pull request merged**